network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/NodeBreakerViewImpl.java

```diff
@@ -125,44 +126,110 @@ public void traverse(int[] nodes, VoltageLevel.NodeBreakerView.TopologyTraverser
+            boolean keepGoing = true;
+            Set<Edge> encounteredEdges = new HashSet<>();
+
+            LinkedList<Integer> vertexToTraverse = new LinkedList<>();
+            vertexToTraverse.offer(node);
+            while (!vertexToTraverse.isEmpty()) {
+                int firstV = vertexToTraverse.getFirst();
+                vertexToTraverse.poll();
+                if (done.contains(firstV)) {
+                    continue;
+                }
+                done.add(firstV);
+
+                Set<Edge> adjacentEdges = graph.edgesOf(firstV);
+
+                for (Edge edge : adjacentEdges) {
+                    if (encounteredEdges.contains(edge)) {
+                        continue;
+                    }
+                    encounteredEdges.add(edge);
+
+                    NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
+                    int node1 = biConnectable.getNode1();
+                    int node2 = biConnectable.getNode2();
+
+                    TraverseResult traverserResult;
+                    if (!done.contains(node1)) {
+                        if (biConnectable instanceof SwitchAttributes) {
+                            traverserResult = traverseSwitch(traverser, biConnectable, node2, node1);
+                        } else if (biConnectable instanceof InternalConnectionAttributes) {
+                            traverserResult = traverser.traverse(node2, null, node1);
+                        } else {
+                            throw new AssertionError();
+                        }
+                        if (traverserResult == TraverseResult.CONTINUE) {
+                            vertexToTraverse.offer(node1);
+                        } else if (traverserResult == TraverseResult.TERMINATE_TRAVERSER) {
+                            keepGoing = false;
+                        }
+                    } else if (!done.contains(node2)) {
+                        if (biConnectable instanceof SwitchAttributes) {
+                            traverserResult = traverseSwitch(traverser, biConnectable, node1, node2);
+                        } else if (biConnectable instanceof InternalConnectionAttributes) {
+                            traverserResult = traverser.traverse(node1, null, node2);
+                        } else {
+                            throw new AssertionError();
+                        }
+                        if (traverserResult == TraverseResult.CONTINUE) {
+                            vertexToTraverse.offer(node2);
+                        } else if (traverserResult == TraverseResult.TERMINATE_TRAVERSER) {
+                            keepGoing = false;
+                        }
+                    }
+                    if (!keepGoing) {
+                        break;
+                    }
+                }
+                if (!keepGoing) {
+                    break;
+                }
+            }
+            return keepGoing;
```

Comment:
Something like this seems to work for me on tests but I haven't tested on a real network only on the tests implemented. 
```
    boolean traverseFromNode(int node, TraversalType traversalType, VoltageLevel.NodeBreakerView.TopologyTraverser traverser) {
        Graph<Integer, Edge> graph = NodeBreakerTopology.INSTANCE.buildGraph(index, getVoltageLevelResource(), true, true);
        if (traversalType == TraversalType.DEPTH_FIRST) {
            Set<Integer> visitedNodes = new HashSet<>();
            return traverseFromNodeDFS(graph, node, traverser, visitedNodes);
        } else {
            return traverseFromNodeBFS(graph, node, traverser);
        }
    }

    private boolean traverseFromNodeBFS(Graph<Integer, Edge> graph, int node, TopologyTraverser traverser) {
        Set<Integer> visitedNodes = new HashSet<>();
        LinkedList<Integer> nodeQueue = new LinkedList<>();
        nodeQueue.offer(node);

        while (!nodeQueue.isEmpty()) {
            int currentNode = nodeQueue.poll();

            for (Edge edge : graph.edgesOf(currentNode)) {
                NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
                int nextNode = biConnectable.getNode1() == currentNode ? biConnectable.getNode2() : biConnectable.getNode1();
                if (visitedNodes.contains(nextNode)) {
                    continue;
                }
                visitedNodes.add(nextNode);

                TraverseResult traverserResult;
                if (biConnectable instanceof SwitchAttributes) {
                    traverserResult = traverseSwitch(traverser, biConnectable, currentNode, nextNode);
                } else if (biConnectable instanceof InternalConnectionAttributes) {
                    traverserResult = traverser.traverse(currentNode, null, nextNode);
                } else {
                    throw new AssertionError();
                }

                if (traverserResult == TraverseResult.CONTINUE) {
                    nodeQueue.offer(nextNode);
                } else if (traverserResult == TraverseResult.TERMINATE_TRAVERSER) {
                    return false;
                }
            }
        }
        return true;
    }

    private boolean traverseFromNodeDFS(Graph<Integer, Edge> graph, int node,
                                        TopologyTraverser traverser, Set<Integer> visitedNodes) {
        if (visitedNodes.contains(node)) {
            return true;
        }
        visitedNodes.add(node);

        for (Edge edge : graph.edgesOf(node)) {
            NodeBreakerBiConnectable biConnectable = edge.getBiConnectable();
            int nextNode = biConnectable.getNode1() == node ? biConnectable.getNode2() : biConnectable.getNode1();
            TraverseResult result;
            if (visitedNodes.contains(nextNode)) {
                continue;
            }
            if (biConnectable instanceof SwitchAttributes) {
                result = traverseSwitch(traverser, biConnectable, node, nextNode);
            } else if (biConnectable instanceof InternalConnectionAttributes) {
                result = traverser.traverse(node, null, nextNode);
            } else {
                throw new AssertionError();
            }
            if (result == TraverseResult.CONTINUE) {
                if (!traverseFromNodeDFS(graph, nextNode, traverser, visitedNodes)) {
                    return false;
                }
            } else if (result == TraverseResult.TERMINATE_TRAVERSER) {
                return false;
            }
        }
        return true;
    }
```